### PR TITLE
Apply event sourcing to sequence flows

### LINKED ISSUE #6450: Remove the BPMN sequence flow processor
**Description**

After https://github.com/zeebe-io/zeebe/issues/6190, the sequence flow processor can be removed. Any work done on processing the record, can be moved to the processing of `SEQUENCE_FLOW_TAKEN` source command. 

Reference:
* [ZEP](https://github.com/zeebe-io/enhancements/blob/master/ZEP004-wf-stream-processing.md#bpmn-processing)

### PR DESCRIPTION
**Pull request by korthout**
## Description

Since the only record that is processed for sequence flows is an event,
simply remove the processor and move its logic elsewhere.

The sequence flow processor kept track of taken incoming sequence flows
of a joining parallel gateway. This gateway should only be activated
once all incoming sequence flows have been taken. 

- Keeping track of which flows are taken is now tracked by the sequence flow
taken applier.
- Checking whether or not to activate the parallel gateway is a done in
the state transition behavior - uses SequenceFlowAnalyzer
- Cleanup of the state of taken incoming sequence flows is done on activating
the target element (parallel gateway) - uses SequenceFlowAnalyzer

## Related issues

closes #6450 

## Definition of Done

_Not all items need to be done depending on the issue and the pull request._

Code changes:
* [ ] The changes are backwards compatibility with previous versions
* [ ] If it fixes a bug then PRs are created to [backport](https://github.com/zeebe-io/zeebe/compare/stable/0.24...develop?expand=1&template=backport_template.md&title=[Backport%200.24]) the fix to the last two minor versions. You can trigger a backport by assigning labels (e.g. `backport stable/0.25`) to the PR, in case that fails you need to create backports manually.

Testing:
* [x] There are unit/integration tests that verify all acceptance criterias of the issue
* [ ] New tests are written to ensure backwards compatibility with further versions
* [ ] The behavior is tested manually
* [ ] The change has been verified by a QA run
* [ ] The impact of the changes is verified by a benchmark 

Documentation: 
* [ ] The documentation is updated (e.g. BPMN reference, configuration, examples, get-started guides, etc.)
* [ ] New content is added to the [release announcement](https://drive.google.com/drive/u/0/folders/1DTIeswnEEq-NggJ25rm2BsDjcCQpDape)

### CODE REVIEW

File: engine/src/main/java/io/zeebe/engine/state/analyzers/SequenceFlowAnalyzer.java [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+
+  /**
+   * Uses deferred records to determine which incoming sequence flows have been taken. Note that a
+   * sequence flow may have been taken more than once.
+   *
+   * @param flowScopeKey flow scope key of the element whose taken incoming sequence flows we want
+   *     to determine. The deferred records must be stored at this key's scope
+   * @param element The target element for the incoming sequence flows
+   * @return a map of deferred records of incoming sequence flows that have been taken, grouped by
+   *     the elementId of the sequence flows
```
Comment (Zelldon) [inline engine/src/main/java/io/zeebe/engine/state/analyzers/SequenceFlowAnalyzer.java:29]: Maybe we can improve the description a bit, example:
```
   A map which consist of: 
     key:  (taken sequence flow) -> value: (list of deferred records, equal to how many tokens took this sequence flow)
     They keyset size is equal to how many sequence flows are already taken to the given target.
```

File: engine/src/main/java/io/zeebe/engine/state/appliers/ProcessInstanceElementActivatingApplier.java [hunk 2]
```diff
@@ -110,6 +115,30 @@ public void applyState(final long elementInstanceKey, final ProcessInstanceRecor
...
+        (sequenceFlow, tokens) -> {
+          // before a parallel gateway is activated, each incoming sequence flow of the gateway must
+          // be taken (at least once). if a sequence flow is taken more than once then the redundant
+          // token remains for the next activation of the gateway (Tetris principle)
+          final var firstToken = tokens.get(0);
+          elementInstanceState.removeStoredRecord(
+              value.getFlowScopeKey(), firstToken.getKey(), Purpose.DEFERRED);
+        });
+  }
+
```
Comment (Zelldon) [inline engine/src/main/java/io/zeebe/engine/state/appliers/ProcessInstanceElementActivatingApplier.java:141]: Are deferred records are still used elsewhere? IF not then we can just say delete all. Maybe we rename these to awaiting parallel gw records or something. I think anyway we could simplify this, like I wrote above just storing the element id and some key for example.

**Review by ChrisKujawa — APPROVED** [review]
Thanks @korthout good work :+1: :100: 

I just have a remark regarding the deferred records, maybe we can create a follow up to make this simpler or you do it immediately you can decide.

File: engine/src/main/java/io/zeebe/engine/state/appliers/ProcessInstanceSequenceFlowTakenApplier.java [hunk 3]
```diff
@@ -37,5 +47,56 @@ public void applyState(final long key, final ProcessInstanceRecord value) {
...
+            elementInstanceState.removeStoredRecord(
+                value.getFlowScopeKey(), firstToken.getKey(), Purpose.DEFERRED);
+          });
+
+      // finally write the active token changes to state
+      elementInstanceState.updateInstance(flowScopeInstance);
+    }
+  }
+
+  // copied from BpmnStateTransitionBehavior
```
[the remark this answers is not in the document]
Comment (Zelldon) [reply engine/src/main/java/io/zeebe/engine/state/appliers/ProcessInstanceSequenceFlowTakenApplier.java]: It is weird to me to review commit by commit :see_no_evil: please ignore it. I then reviewed all changes again.

[the remark this answers is not in the document]
Comment (korthout) [reply engine/src/main/java/io/zeebe/engine/state/appliers/ProcessInstanceSequenceFlowTakenApplier.java]: Let's talk about how to improve this

File: engine/src/main/java/io/zeebe/engine/state/analyzers/SequenceFlowAnalyzer.java [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+
+  private final ElementInstanceState elementInstanceState;
+
+  public SequenceFlowAnalyzer(final ElementInstanceState elementInstanceState) {
+    this.elementInstanceState = elementInstanceState;
+  }
+
+  /**
+   * Uses deferred records to determine which incoming sequence flows have been taken. Note that a
+   * sequence flow may have been taken more than once.
```
[the remark this answers is not in the document]
Comment (korthout) [reply engine/src/main/java/io/zeebe/engine/state/analyzers/SequenceFlowAnalyzer.java:23]: How would that work, would we need a new "zeebe" column family for that? I am assuming that we can't just use `elementInstanceState` for it, but perhaps that's not true

File: engine/src/main/java/io/zeebe/engine/state/appliers/ProcessInstanceElementActivatingApplier.java (hunk 2 above)
Comment (korthout) [reply engine/src/main/java/io/zeebe/engine/state/appliers/ProcessInstanceElementActivatingApplier.java:141]: Seems to still be used in multi-instance body processor and event sub-process processor. Let's add an additional issue to follow-up on removing deferred records

Comment (Zelldon) [reply engine/src/main/java/io/zeebe/engine/state/appliers/ProcessInstanceElementActivatingApplier.java:141]: sounds good

Comment (korthout) [reply engine/src/main/java/io/zeebe/engine/state/appliers/ProcessInstanceElementActivatingApplier.java:141]: Created: https://github.com/camunda-cloud/zeebe/issues/6832

File: engine/src/main/java/io/zeebe/engine/state/analyzers/SequenceFlowAnalyzer.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (korthout) [reply engine/src/main/java/io/zeebe/engine/state/analyzers/SequenceFlowAnalyzer.java]: Created #6831

**Comment by korthout** [comment]
bors merge

**Comment by korthout** [comment]
bors merge

**Comment by korthout** [comment]
not related flaky test

bors retry

**Pull request merged**